destination/writer/writer.go

```diff
@@ -0,0 +1,293 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/conduitio-labs/conduit-connector-clickhouse/columntypes"
+	"github.com/conduitio-labs/conduit-connector-clickhouse/repository"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+)
+
+const (
+	// metadata key with the name of the table.
+	metadataTable = "clickhouse.table"
+
+	// query patterns.
+	insertFmt = "INSERT INTO %s (%s) VALUES (%s)"
+	updateFmt = "ALTER TABLE %s UPDATE %s WHERE %s"
+	deleteFmt = "ALTER TABLE %s DELETE WHERE %s"
+
+	coma        = ","
+	placeholder = "?"
+	operatorAnd = " AND "
+)
+
+// Writer implements a writer logic for ClickHouse destination.
+type Writer struct {
+	repo           *repository.ClickHouse
+	table          string
+	primaryColumns []string
+	columnTypes    map[string]string
+}
+
+// Params is an incoming params for the New function.
+type Params struct {
+	Repo           *repository.ClickHouse
+	Table          string
+	PrimaryColumns []string
+}
+
+// New creates new instance of the Writer.
+func New(ctx context.Context, params Params) (*Writer, error) {
+	writer := &Writer{
+		repo:           params.Repo,
+		table:          params.Table,
+		primaryColumns: params.PrimaryColumns,
+	}
+
+	columnTypes, err := columntypes.GetColumnTypes(ctx, writer.repo, writer.table)
+	if err != nil {
+		return nil, fmt.Errorf("get column types: %w", err)
+	}
+
+	writer.columnTypes = columnTypes
+
+	return writer, nil
+}
+
+// Write writes a sdk.Record into a Destination.
+func (w *Writer) Write(ctx context.Context, record sdk.Record) error {
+	switch record.Operation {
+	case sdk.OperationUpdate:
+		return w.update(ctx, record)
+	case sdk.OperationDelete:
+		return w.delete(ctx, record)
+	default:
+		return w.insert(ctx, record)
+	}
+}
+
+// inserts a record.
+func (w *Writer) insert(ctx context.Context, record sdk.Record) error {
+	tableName := w.getTableName(record.Metadata)
+
+	payload, err := w.structurizeData(record.Payload.After)
+	if err != nil {
+		return fmt.Errorf("structurize payload: %w", err)
+	}
+
+	// if payload is empty return empty payload error
+	if payload == nil {
+		return errNoPayload
+	}
+
+	payload, err = columntypes.ConvertStructureData(w.columnTypes, payload)
+	if err != nil {
+		return fmt.Errorf("convert structure data: %w", err)
+	}
+
+	key, err := w.structurizeData(record.Key)
+	if err != nil {
+		// if the key is not structured, we simply ignore it
+		// we'll try to insert just a payload in this case
+		sdk.Logger(ctx).Debug().Msgf("structurize key during insert: %v", err)
+	}
+
+	keyColumns, err := w.getKeyColumns(key)
+	if err != nil {
+		return fmt.Errorf("get key columns: %w", err)
+	}
+
+	// if the record doesn't contain some key column, insert it
+	for i := range keyColumns {
```

Comment:
Do we need do it now? Maybe we can simplify and insert without check key

Reply:
No, I've rewritten it.